providers/ms_entra_id.go

```diff
@@ -125,29 +127,79 @@ func (p *MicrosoftEntraIDProvider) redeemWithFederatedToken(ctx context.Context,
 	params.Add("code", code)
 	params.Add("grant_type", "authorization_code")
 
-	// perform exchange
-	resp := requests.New(p.RedeemURL.String()).
-		WithContext(ctx).
-		WithMethod("POST").
-		WithBody(bytes.NewBufferString(params.Encode())).
-		SetHeader("Content-Type", "application/x-www-form-urlencoded").
-		Do()
+	token, err := p.fetchToken(ctx, params)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching token: %w", err)
+	}
 
-	// prepare token of type *oauth2.Token
-	var token *oauth2.Token
-	var rawResponse interface{}
+	return p.OIDCProvider.createSession(ctx, token, false)
+}
 
-	body := resp.Body()
-	if err := json.Unmarshal(body, &rawResponse); err != nil {
-		return nil, err
+// RefreshSession uses the RefreshToken to fetch new Access and ID Tokens
+func (p *MicrosoftEntraIDProvider) RefreshSession(ctx context.Context, s *sessions.SessionState) (bool, error) {
+	if s == nil || s.RefreshToken == "" {
+		return false, nil
 	}
 
-	if err := json.Unmarshal(body, &token); err != nil {
-		return nil, err
+	var err error
+	ctx = oidc.ClientContext(ctx, requests.DefaultHTTPClient)
+	if p.federatedTokenAuth {
+		err = p.redeemRefreshTokenWithFederatedToken(ctx, s)
+	} else {
+		err = p.redeemRefreshToken(ctx, s)
 	}
 
-	// create session using new token and generic OIDC provider
-	return p.OIDCProvider.createSession(ctx, token.WithExtra(rawResponse), false)
+	if err != nil {
+		return false, fmt.Errorf("unable to redeem refresh token: %v", err)
+	}
+
+	return true, nil
+}
+
+// redeemRefreshTokenWithFederatedToken uses a RefreshToken and federacted credentials with the RedeemURL to refresh the
+// Access Token and (probably) the ID Token.
+func (p *MicrosoftEntraIDProvider) redeemRefreshTokenWithFederatedToken(ctx context.Context, s *sessions.SessionState) error {
+	federatedTokenPath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
+	federatedToken, err := os.ReadFile(federatedTokenPath)
+	if err != nil {
+		return fmt.Errorf("error reading federated token file %s: %s", federatedTokenPath, err)
+	}
+
+	params := url.Values{}
+	params.Add("client_id", p.ClientID)
+	params.Add("client_assertion", string(federatedToken))
+	params.Add("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+	params.Add("refresh_token", s.RefreshToken)
+	params.Add("grant_type", "refresh_token")
+	params.Add("expiry", time.Now().Add(-time.Hour).Format(time.RFC3339))
+
+	token, err := p.fetchToken(ctx, params)
+	if err != nil {
+		return fmt.Errorf("error fetching token: %w", err)
+	}
+
+	newSession, err := p.OIDCProvider.createSession(ctx, token, true)
+	if err != nil {
+		return fmt.Errorf("unable create new session state from response: %v", err)
+	}
+
+	// It's possible that if the refresh token isn't in the token response the
+	// session will not contain an id token.
+	// If it doesn't it's probably better to retain the old one
```

Comment:
This is in accordance with the OIDC spec. The ID token does not need to be refresh only the access token. Therefore please remove this vague phrasing.

Better to shortly explain and link:

https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse

Upon successful validation of the Refresh Token, the response body is the Token Response of [Section 3.1.3.3](https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse) except that it might not contain an id_token.

Reply:
Replaced with:
```
	// Optionally update the ID Token if it's returned
	// ref. https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse
```